Dashboard: Add a feature that creates a table panel when a spreadsheet file is dropped on the dashboard.

oscarkilhed opened the pull request:
**What is this feature?**

![draganddropfeature](https://user-images.githubusercontent.com/468940/216069663-a0e0817f-6cd9-4e99-848c-a8f918832d7b.gif)

This feature lets users drag spreadsheet files to a dashboard to create a table panel with a spreadsheet query set up to the Grafana datasource.

**Why do we need this feature?**

It lets users quickly get data they have locally into Grafana.

**Who is this feature for?**

New users of Grafana that just wants to try it out together with some data they have, without setting up a datasource.

Existing users that have data locally that they want to be able to view together with data from their set up datasources.

oscarkilhed committed:
drag files to dashboard

oscarkilhed committed:
update styling

oscarkilhed committed:
add a

oscarkilhed committed:
use file name as panel title

oscarkilhed committed:
add file size limitation, file type limitation and error handling

oscarkilhed committed:
Merge branch 'main' into oscark/drag-to-dashboard

ryantxu commented:
This looks great!  It would be even better if the drop would accept a dashboard JSON and then ask to import it... eg https://github.com/grafana/grafana/pull/61017

In addition to drag+drop CSV/text, we should also support general JSON.  (use `toDataFrame(...)`)

codeincarnate commented on public/app/features/dashboard/containers/DashboardPage.tsx:
```diff
@@ -97,6 +111,69 @@ export class UnthemedDashboardPage extends PureComponent<Props, State> {
 
   private forceRouteReloadCounter = 0;
   state: State = this.getCleanState();
+  maxFileSize = 500000;
```
Would be good to pull this into a module level constant (i.e. MAX_FILESIZE or similar)

codeincarnate commented on public/app/features/dashboard/containers/DashboardPage.tsx:
```diff
@@ -97,6 +111,69 @@ export class UnthemedDashboardPage extends PureComponent<Props, State> {
+  maxFileSize = 500000;
+
+  onFileDrop = (acceptedFiles: File[], fileRejections: FileRejection[], event: DropEvent) => {
+    console.log(acceptedFiles);
```
console.log 👀

codeincarnate commented on public/app/features/dashboard/containers/DashboardPage.tsx:
```diff
@@ -97,6 +111,69 @@ export class UnthemedDashboardPage extends PureComponent<Props, State> {
+            type: 'grafana',
+            uid: 'grafana',
+          };
+          this.props.dashboard?.addPanel({
```
Seems like the creation of this literal should be extracted into a function or similar. Seems a little bit awkward this way

oscarkilhed replied:
I agree. IDK if it's possible to get this from the plugin.json for the grafana ds in any way. Would be nice.

codeincarnate commented on public/app/features/dashboard/containers/DashboardPage.tsx:
```diff
@@ -378,21 +455,53 @@ export class UnthemedDashboardPage extends PureComponent<Props, State> {
+            <DropZone
+              onDrop={this.onFileDrop}
+              accept={{
+                'text/plain': ['.csv', '.txt'],
```
Would also be good to have a constant for accepted file types. Something like DND_ACCEPTED_FILE_TYPES or similar

codeincarnate commented on public/app/features/dashboard/containers/DashboardPage.tsx:
```diff
@@ -97,6 +111,69 @@ export class UnthemedDashboardPage extends PureComponent<Props, State> {
+            const formattedSize = getValueFormat('decbytes')(this.maxFileSize);
+            return `File is larger than ${formattedValueToString(formattedSize)}.`;
+          case ErrorCode.FileInvalidType:
+            return 'File must be a spreadsheet.';
```
Would be good to have more guidance in error message such as having a short list of file types. Right now it's hard to tell what exactly will work and what won't

oscarkilhed replied:
Addressed here: https://github.com/grafana/grafana/pull/62688/commits/d38cec40182393edc3b57ca565e5a4069c523c9f#diff-a09c348290db927b350a7af939ca2bca0db0d3151f1999a963d1bf8356fa9c5eR149

codeincarnate commented on public/app/features/dashboard/containers/DashboardPage.tsx:
```diff
@@ -97,6 +111,69 @@ export class UnthemedDashboardPage extends PureComponent<Props, State> {
+        switch (error.code) {
+          case ErrorCode.FileTooLarge:
+            const formattedSize = getValueFormat('decbytes')(this.maxFileSize);
+            return `File is larger than ${formattedValueToString(formattedSize)}.`;
```
Active voice for the error message would be better here. e.g. `File must be less than X size.`

oscarkilhed replied:
https://github.com/grafana/grafana/pull/62688/commits/d38cec40182393edc3b57ca565e5a4069c523c9f#diff-a09c348290db927b350a7af939ca2bca0db0d3151f1999a963d1bf8356fa9c5eR149

Addressed here now

codeincarnate commented:
A couple of more things, the uneven padding (left vs top) becomes really obvious when the dropzone is shown and looks weird.

As for the note about the dropzone message it seems like we should be able to tell cases with 1 file and more than 1 file? Seems like it would also be good to provide more file type guidance here, like show a short list of accepted file types

<img width="894" alt="Screenshot 2023-02-06 at 1 34 54 PM" src="https://user-images.githubusercontent.com/199847/216902286-f2ddf495-94ab-4251-9f3b-7990717c1966.png">

oscarkilhed commented:
> As for the note about the dropzone message it seems like we should be able to tell cases with 1 file and more than 1 file?

Most browsers won't divulge any info about the dragged files until they are dropped.

oscarkilhed committed:
Refactor file parsing for code sharing

move accepted types and max size to file-import constants
show which file types are allowed in file type error

oscarkilhed committed:
Merge branch 'main' into oscark/drag-to-dashboard

oscarkilhed commented:
@ryantxu the last commit added support for dropping json files that contain dataframes. 

However, adding support for importing dropped dashboard json files comes with a whole host of UX questions.

oscarkilhed committed:
update codeowners

ryantxu commented on public/app/features/dataframe-import/utils.ts:
```diff
@@ -0,0 +1,50 @@
+  return `${fileExtentions.slice(0, -1).join(', ')} or ${fileExtentions.slice(-1)}`;
+}
+
+export function filesToDataframes(files: File[]): Observable<FileImportResult> {
```
Rather than supporting multiple async files, can we limit this to a single file at a time?  I don't think we want dropping a pile of csv files to generate a pile of table panels -- at least not to start!

ryantxu replied:
looks like the drop component enforces that, so maybe this is fine 🤷

codeincarnate reviewed (approved):
Thanks for the updates Oscar. LGTM! 🚀

zoltanbedi reviewed (commented):
Can you please add some docs around this?

oscarkilhed committed:
Adjust max size to 1mb

oscarkilhed committed:
Merge branch 'main' into oscark/drag-to-dashboard
